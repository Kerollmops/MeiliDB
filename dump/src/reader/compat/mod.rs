// pub mod v2;
// pub mod v3;
// pub mod v4;
use crate::Result;

use self::{
    v4_to_v5::CompatV4ToV5,
    v5_to_v6::{CompatIndexV5ToV6, CompatV5ToV6},
};

use super::{
    v5::V5Reader,
    v6::{self, V6IndexReader, V6Reader},
    Document, UpdateFile,
};

pub mod v2_to_v3;
pub mod v3_to_v4;
pub mod v4_to_v5;
pub mod v5_to_v6;

pub enum Compat {
    Current(V6Reader),
    Compat(CompatV5ToV6),
}

impl Compat {
    pub fn version(&self) -> crate::Version {
        match self {
            Compat::Current(current) => current.version(),
            Compat::Compat(compat) => compat.version(),
        }
    }

    pub fn date(&self) -> Option<time::OffsetDateTime> {
        match self {
            Compat::Current(current) => current.date(),
            Compat::Compat(compat) => compat.date(),
        }
    }

    pub fn instance_uid(&self) -> Result<Option<uuid::Uuid>> {
        match self {
            Compat::Current(current) => current.instance_uid(),
            Compat::Compat(compat) => compat.instance_uid(),
        }
    }

    pub fn indexes(&self) -> Result<Box<dyn Iterator<Item = Result<CompatIndex>> + '_>> {
        match self {
            Compat::Current(current) => {
                let indexes = Box::new(current.indexes()?.map(|res| res.map(CompatIndex::from)))
                    as Box<dyn Iterator<Item = Result<CompatIndex>> + '_>;
                Ok(indexes)
            }
            Compat::Compat(compat) => {
                let indexes = Box::new(compat.indexes()?.map(|res| res.map(CompatIndex::from)))
                    as Box<dyn Iterator<Item = Result<CompatIndex>> + '_>;
                Ok(indexes)
            }
        }
    }

    pub fn tasks(
        &mut self,
    ) -> Box<dyn Iterator<Item = Result<(v6::Task, Option<Box<UpdateFile>>)>> + '_> {
        match self {
            Compat::Current(current) => current.tasks(),
            Compat::Compat(compat) => compat.tasks(),
        }
    }

    pub fn keys(&mut self) -> Box<dyn Iterator<Item = Result<v6::Key>> + '_> {
        match self {
            Compat::Current(current) => current.keys(),
            Compat::Compat(compat) => compat.keys(),
        }
    }
}

impl From<V6Reader> for Compat {
    fn from(value: V6Reader) -> Self {
        Compat::Current(value)
    }
}

impl From<CompatV5ToV6> for Compat {
    fn from(value: CompatV5ToV6) -> Self {
        Compat::Compat(value)
    }
}

impl From<V5Reader> for Compat {
    fn from(value: V5Reader) -> Self {
        Compat::Compat(value.to_v6())
    }
}

impl From<CompatV4ToV5> for Compat {
    fn from(value: CompatV4ToV5) -> Self {
        Compat::Compat(value.to_v6())
    }
}

pub enum CompatIndex {
    Current(v6::V6IndexReader),
    Compat(CompatIndexV5ToV6),
}

impl CompatIndex {
    pub fn new_v6(v6: v6::V6IndexReader) -> CompatIndex {
        CompatIndex::Current(v6)
    }

    pub fn metadata(&self) -> &crate::IndexMetadata {
        match self {
            CompatIndex::Current(v6) => v6.metadata(),
            CompatIndex::Compat(compat) => compat.metadata(),
        }
    }

    pub fn documents(&mut self) -> Result<Box<dyn Iterator<Item = Result<Document>> + '_>> {
        match self {
            CompatIndex::Current(v6) => v6
                .documents()
                .map(|iter| Box::new(iter) as Box<dyn Iterator<Item = Result<Document>> + '_>),
            CompatIndex::Compat(compat) => compat
                .documents()
                .map(|iter| Box::new(iter) as Box<dyn Iterator<Item = Result<Document>> + '_>),
        }
    }

    pub fn settings(&mut self) -> Result<v6::Settings<v6::Checked>> {
        match self {
            CompatIndex::Current(v6) => v6.settings(),
            CompatIndex::Compat(compat) => compat.settings(),
        }
    }
}

impl From<V6IndexReader> for CompatIndex {
    fn from(value: V6IndexReader) -> Self {
        CompatIndex::Current(value)
    }
}

impl From<CompatIndexV5ToV6> for CompatIndex {
    fn from(value: CompatIndexV5ToV6) -> Self {
        CompatIndex::Compat(value)
    }
}
