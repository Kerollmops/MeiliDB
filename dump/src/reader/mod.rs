use std::io::Read;
use std::{fs::File, io::BufReader};

use flate2::bufread::GzDecoder;

use serde::Deserialize;

use tempfile::TempDir;
use time::OffsetDateTime;
use uuid::Uuid;

// use crate::reader::compat::Compat;
use crate::{IndexMetadata, Result, Version};

use self::compat::Compat;

// use self::loaders::{v2, v3, v4, v5};

// pub mod error;
mod compat;
// mod loaders;
// mod v1;
pub(self) mod v2;
pub(self) mod v3;
pub(self) mod v4;
pub(self) mod v5;
pub(self) mod v6;

pub type Document = serde_json::Map<String, serde_json::Value>;
pub type UpdateFile = dyn Iterator<Item = Result<Document>>;

pub fn open(dump: impl Read) -> Result<Compat> {
    let path = TempDir::new()?;
    let mut dump = BufReader::new(dump);
    let gz = GzDecoder::new(&mut dump);
    let mut archive = tar::Archive::new(gz);
    archive.unpack(path.path())?;

    #[derive(Deserialize)]
    #[serde(rename_all = "camelCase")]
    struct MetadataVersion {
        pub dump_version: Version,
    }
    let mut meta_file = File::open(path.path().join("metadata.json"))?;
    let MetadataVersion { dump_version } = serde_json::from_reader(&mut meta_file)?;

    match dump_version {
        // Version::V1 => Ok(Box::new(v1::Reader::open(path)?)),
        Version::V1 => todo!(),
        Version::V2 => todo!(),
        Version::V3 => Ok(v3::V3Reader::open(path)?.to_v4().to_v5().to_v6().into()),
        Version::V4 => Ok(v4::V4Reader::open(path)?.to_v5().to_v6().into()),
        Version::V5 => Ok(v5::V5Reader::open(path)?.to_v6().into()),
        Version::V6 => Ok(v6::V6Reader::open(path)?.into()),
    }
}

pub trait DumpReader {
    /// Return the version of the dump.
    fn version(&self) -> Version;

    /// Return at which date the dump was created if there was one.
    fn date(&self) -> Option<OffsetDateTime>;

    /// Return the instance-uid if there was one.
    fn instance_uid(&self) -> Result<Option<Uuid>>;

    /// Return an iterator over each indexes.
    fn indexes(&self) -> Result<Box<dyn Iterator<Item = Result<Box<dyn IndexReader + '_>>> + '_>>;

    /// Return all the tasks in the dump with a possible update file.
    fn tasks(
        &mut self,
    ) -> Box<dyn Iterator<Item = Result<(v6::Task, Option<v6::UpdateFile>)>> + '_>;

    /// Return all the keys.
    fn keys(&mut self) -> Box<dyn Iterator<Item = Result<v6::Key>> + '_>;
}

pub trait IndexReader {
    fn metadata(&self) -> &IndexMetadata;
    fn documents(&mut self) -> Result<Box<dyn Iterator<Item = Result<v6::Document>> + '_>>;
    fn settings(&mut self) -> Result<v6::Settings<v6::Checked>>;
}
