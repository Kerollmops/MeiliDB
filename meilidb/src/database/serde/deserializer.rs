use std::error::Error;
use std::ops::Deref;
use std::fmt;

use rocksdb::rocksdb::{DB, Snapshot, SeekKey};
use rocksdb::rocksdb_options::ReadOptions;
use serde::forward_to_deserialize_any;
use serde::de::value::MapDeserializer;
use serde::de::{self, Visitor, IntoDeserializer};

use crate::database::document_key::{DocumentKey, DocumentKeyAttr};
use crate::database::schema::Schema;
use meilidb_core::DocumentId;

pub struct Deserializer<'a, D>
where D: Deref<Target=DB>
{
    snapshot: &'a Snapshot<D>,
    schema: &'a Schema,
    document_id: DocumentId,
}

impl<'a, D> Deserializer<'a, D>
where D: Deref<Target=DB>
{
    pub fn new(snapshot: &'a Snapshot<D>, schema: &'a Schema, doc: DocumentId) -> Self {
        Deserializer { snapshot, schema, document_id: doc }
    }
}

impl<'de, 'a, 'b, D> de::Deserializer<'de> for &'b mut Deserializer<'a, D>
where D: Deref<Target=DB>
{
    type Error = DeserializerError;

    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where V: Visitor<'de>
    {
        self.deserialize_map(visitor)
    }

    forward_to_deserialize_any! {
        bool u8 u16 u32 u64 i8 i16 i32 i64 f32 f64 char str string unit seq
        bytes byte_buf unit_struct tuple_struct
        identifier tuple ignored_any option newtype_struct enum struct
    }

    fn deserialize_map<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where V: Visitor<'de>
    {
        let mut options = ReadOptions::new();
        let lower = DocumentKey::new(self.document_id);
        let upper = lower.with_attribute_max();
        options.set_iterate_lower_bound(lower.as_ref());
        options.set_iterate_upper_bound(upper.as_ref());

        let mut iter = self.snapshot.iter_opt(options);
        iter.seek(SeekKey::Start);

        if iter.kv().is_none() {
            // FIXME return an error
        }

        let iter = iter.map(|(key, value)| {
            // retrieve the schema attribute name
            // from the schema attribute number
            let document_key_attr = DocumentKeyAttr::from_bytes(&key);
            let schema_attr = document_key_attr.attribute();
            let attribute_name = self.schema.attribute_name(schema_attr);
            (attribute_name, Value(value))
        });

        let map_deserializer = MapDeserializer::new(iter);
        visitor.visit_map(map_deserializer)
    }
}

struct Value(Vec<u8>);

impl<'de> IntoDeserializer<'de, DeserializerError> for Value {
    type Deserializer = Self;

    fn into_deserializer(self) -> Self::Deserializer {
        self
    }
}

macro_rules! forward_to_bincode_values {
    ($($ty:ident => $de_method:ident,)*) => {
        $(
            fn $de_method<V>(self, visitor: V) -> Result<V::Value, Self::Error>
                where V: de::Visitor<'de>
            {
                match bincode::deserialize::<$ty>(&self.0) {
                    Ok(val) => val.into_deserializer().$de_method(visitor),
                    Err(e) => Err(de::Error::custom(e)),
                }
            }
        )*
    }
}

impl<'de, 'a> de::Deserializer<'de> for Value {
    type Error = DeserializerError;

    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where V: Visitor<'de>
    {
        self.0.into_deserializer().deserialize_any(visitor)
    }

    fn deserialize_str<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where V: Visitor<'de>
    {
        self.deserialize_string(visitor)
    }

    fn deserialize_string<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where V: Visitor<'de>
    {
        match bincode::deserialize::<String>(&self.0) {
            Ok(val) => val.into_deserializer().deserialize_string(visitor),
            Err(e) => Err(de::Error::custom(e)),
        }
    }

    fn deserialize_bytes<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where V: Visitor<'de>
    {
        self.deserialize_byte_buf(visitor)
    }

    fn deserialize_byte_buf<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where V: Visitor<'de>
    {
        match bincode::deserialize::<Vec<u8>>(&self.0) {
            Ok(val) => val.into_deserializer().deserialize_byte_buf(visitor),
            Err(e) => Err(de::Error::custom(e)),
        }
    }

    forward_to_bincode_values! {
        char => deserialize_char,
        bool => deserialize_bool,

        u8  => deserialize_u8,
        u16 => deserialize_u16,
        u32 => deserialize_u32,
        u64 => deserialize_u64,

        i8  => deserialize_i8,
        i16 => deserialize_i16,
        i32 => deserialize_i32,
        i64 => deserialize_i64,

        f32 => deserialize_f32,
        f64 => deserialize_f64,
    }

    forward_to_deserialize_any! {
        unit seq map
        unit_struct tuple_struct
        identifier tuple ignored_any option newtype_struct enum struct
    }
}

#[derive(Debug)]
pub enum DeserializerError {
    Custom(String),
}

impl de::Error for DeserializerError {
    fn custom<T: fmt::Display>(msg: T) -> Self {
        DeserializerError::Custom(msg.to_string())
    }
}

impl fmt::Display for DeserializerError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            DeserializerError::Custom(s) => f.write_str(&s),
        }
    }
}

impl Error for DeserializerError {}
