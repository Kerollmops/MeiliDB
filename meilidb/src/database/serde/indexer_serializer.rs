use std::collections::HashSet;

use serde::Serialize;
use serde::ser;
use meilidb_core::{DocumentId, DocIndex};
use meilidb_tokenizer::{TokenizerBuilder, Token, is_cjk};

use crate::database::update::DocumentUpdate;
use crate::database::serde::SerializerError;
use crate::database::schema::SchemaAttr;

pub struct IndexerSerializer<'a, 'b, B> {
    pub tokenizer_builder: &'a B,
    pub update: &'a mut DocumentUpdate<'b>,
    pub document_id: DocumentId,
    pub attribute: SchemaAttr,
    pub stop_words: &'a HashSet<String>,
}

impl<'a, 'b, B> ser::Serializer for IndexerSerializer<'a, 'b, B>
where B: TokenizerBuilder
{
    type Ok = ();
    type Error = SerializerError;
    type SerializeSeq = ser::Impossible<Self::Ok, Self::Error>;
    type SerializeTuple = ser::Impossible<Self::Ok, Self::Error>;
    type SerializeTupleStruct = ser::Impossible<Self::Ok, Self::Error>;
    type SerializeTupleVariant = ser::Impossible<Self::Ok, Self::Error>;
    type SerializeMap = ser::Impossible<Self::Ok, Self::Error>;
    type SerializeStruct = ser::Impossible<Self::Ok, Self::Error>;
    type SerializeStructVariant = ser::Impossible<Self::Ok, Self::Error>;

    forward_to_unserializable_type! {
        bool => serialize_bool,
        char => serialize_char,

        i8  => serialize_i8,
        i16 => serialize_i16,
        i32 => serialize_i32,
        i64 => serialize_i64,

        u8  => serialize_u8,
        u16 => serialize_u16,
        u32 => serialize_u32,
        u64 => serialize_u64,

        f32 => serialize_f32,
        f64 => serialize_f64,
    }

    fn serialize_str(self, v: &str) -> Result<Self::Ok, Self::Error> {
        for token in self.tokenizer_builder.build(v) {
            let Token { word, word_index, char_index } = token;
            let document_id = self.document_id;

            // FIXME must u32::try_from instead
            let attribute = self.attribute.0;
            let word_index = word_index as u16;

            // insert the exact representation
            let word_lower = word.to_lowercase();
            let length = word.chars().count() as u16;

            if self.stop_words.contains(&word_lower) { continue }

            // and the unidecoded lowercased version
            if !word_lower.chars().any(is_cjk) {
                let word_unidecoded = unidecode::unidecode(word).to_lowercase();
                let word_unidecoded = word_unidecoded.trim();
                if word_lower != word_unidecoded {
                    let char_index = char_index as u16;
                    let char_length = length;

                    let doc_index = DocIndex { document_id, attribute, word_index, char_index, char_length };
                    self.update.insert_doc_index(word_unidecoded.as_bytes().to_vec(), doc_index)?;
                }
            }

            let char_index = char_index as u16;
            let char_length = length;

            let doc_index = DocIndex { document_id, attribute, word_index, char_index, char_length };
            self.update.insert_doc_index(word_lower.into_bytes(), doc_index)?;
        }
        Ok(())
    }

    fn serialize_bytes(self, _v: &[u8]) -> Result<Self::Ok, Self::Error> {
        Err(SerializerError::UnserializableType { name: "&[u8]" })
    }

    fn serialize_none(self) -> Result<Self::Ok, Self::Error> {
        Err(SerializerError::UnserializableType { name: "Option" })
    }

    fn serialize_some<T: ?Sized>(self, _value: &T) -> Result<Self::Ok, Self::Error>
    where T: Serialize,
    {
        Err(SerializerError::UnserializableType { name: "Option" })
    }

    fn serialize_unit(self) -> Result<Self::Ok, Self::Error> {
        Err(SerializerError::UnserializableType { name: "()" })
    }

    fn serialize_unit_struct(self, _name: &'static str) -> Result<Self::Ok, Self::Error> {
        Err(SerializerError::UnserializableType { name: "unit struct" })
    }

    fn serialize_unit_variant(
        self,
        _name: &'static str,
        _variant_index: u32,
        _variant: &'static str
    ) -> Result<Self::Ok, Self::Error>
    {
        Err(SerializerError::UnserializableType { name: "unit variant" })
    }

    fn serialize_newtype_struct<T: ?Sized>(
        self,
        _name: &'static str,
        value: &T
    ) -> Result<Self::Ok, Self::Error>
    where T: Serialize,
    {
        value.serialize(self)
    }

    fn serialize_newtype_variant<T: ?Sized>(
        self,
        _name: &'static str,
        _variant_index: u32,
        _variant: &'static str,
        _value: &T
    ) -> Result<Self::Ok, Self::Error>
    where T: Serialize,
    {
        Err(SerializerError::UnserializableType { name: "newtype variant" })
    }

    fn serialize_seq(self, _len: Option<usize>) -> Result<Self::SerializeSeq, Self::Error> {
        Err(SerializerError::UnserializableType { name: "seq" })
    }

    fn serialize_tuple(self, _len: usize) -> Result<Self::SerializeTuple, Self::Error> {
        Err(SerializerError::UnserializableType { name: "tuple" })
    }

    fn serialize_tuple_struct(
        self,
        _name: &'static str,
        _len: usize
    ) -> Result<Self::SerializeTupleStruct, Self::Error>
    {
        Err(SerializerError::UnserializableType { name: "tuple struct" })
    }

    fn serialize_tuple_variant(
        self,
        _name: &'static str,
        _variant_index: u32,
        _variant: &'static str,
        _len: usize
    ) -> Result<Self::SerializeTupleVariant, Self::Error>
    {
        Err(SerializerError::UnserializableType { name: "tuple variant" })
    }

    fn serialize_map(self, _len: Option<usize>) -> Result<Self::SerializeMap, Self::Error> {
        Err(SerializerError::UnserializableType { name: "map" })
    }

    fn serialize_struct(
        self,
        _name: &'static str,
        _len: usize
    ) -> Result<Self::SerializeStruct, Self::Error>
    {
        Err(SerializerError::UnserializableType { name: "struct" })
    }

    fn serialize_struct_variant(
        self,
        _name: &'static str,
        _variant_index: u32,
        _variant: &'static str,
        _len: usize
    ) -> Result<Self::SerializeStructVariant, Self::Error>
    {
        Err(SerializerError::UnserializableType { name: "struct variant" })
    }
}
