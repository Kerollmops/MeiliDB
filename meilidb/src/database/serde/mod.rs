use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};
use std::error::Error;
use std::fmt;

use serde::ser;

macro_rules! forward_to_unserializable_type {
    ($($ty:ident => $se_method:ident,)*) => {
        $(
            fn $se_method(self, _v: $ty) -> Result<Self::Ok, Self::Error> {
                Err(SerializerError::UnserializableType { name: "$ty" })
            }
        )*
    }
}

pub mod find_id;
pub mod key_to_string;
pub mod value_to_number;
pub mod serializer;
pub mod indexer_serializer;
pub mod deserializer;

pub fn calculate_hash<T: Hash>(t: &T) -> u64 {
    let mut s = DefaultHasher::new();
    t.hash(&mut s);
    s.finish()
}

#[derive(Debug)]
pub enum SerializerError {
    DocumentIdNotFound,
    UnserializableType { name: &'static str },
    Custom(String),
}

impl ser::Error for SerializerError {
    fn custom<T: fmt::Display>(msg: T) -> Self {
        SerializerError::Custom(msg.to_string())
    }
}

impl fmt::Display for SerializerError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            SerializerError::DocumentIdNotFound => {
                write!(f, "serialized document does not have an id according to the schema")
            }
            SerializerError::UnserializableType { name } => {
                write!(f, "Only struct and map types are considered valid documents and
                           can be serialized, not {} types directly.", name)
            },
            SerializerError::Custom(s) => f.write_str(&s),
        }
    }
}

impl Error for SerializerError {}

impl From<String> for SerializerError {
    fn from(value: String) -> SerializerError {
        SerializerError::Custom(value)
    }
}
