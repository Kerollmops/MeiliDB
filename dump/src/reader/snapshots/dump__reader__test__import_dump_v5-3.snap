---
source: dump/src/reader/mod.rs
expression: tasks
---
[
  {
    "uid": 21,
    "indexUid": null,
    "status": "enqueued",
    "type": "dumpExport",
    "details": {
      "dumpUid": "20221004-155510279"
    },
    "enqueuedAt": "2022-10-04T15:55:10.281165892Z",
    "startedAt": "2022-10-04T15:55:10.340507253Z"
  },
  {
    "uid": 20,
    "indexUid": "movies_2",
    "status": "enqueued",
    "type": {
      "documentImport": {
        "method": "ReplaceDocuments",
        "allow_index_creation": true
      }
    },
    "details": {
      "receivedDocuments": 200,
      "indexedDocuments": 0
    },
    "enqueuedAt": "2022-10-04T15:55:10.272276202Z"
  },
  {
    "uid": 19,
    "indexUid": "movies",
    "status": "succeeded",
    "type": {
      "documentImport": {
        "method": "ReplaceDocuments",
        "allow_index_creation": true
      }
    },
    "details": {
      "receivedDocuments": 100,
      "indexedDocuments": 100
    },
    "duration": "PT0.062340221S",
    "enqueuedAt": "2022-10-04T15:55:10.259722791Z",
    "startedAt": "2022-10-04T15:55:10.273278199Z",
    "finishedAt": "2022-10-04T15:55:10.33561842Z"
  },
  {
    "uid": 18,
    "indexUid": "dnd_spells",
    "status": "succeeded",
    "type": {
      "documentImport": {
        "method": "ReplaceDocuments",
        "allow_index_creation": true
      }
    },
    "details": {
      "receivedDocuments": 10,
      "indexedDocuments": 10
    },
    "duration": "PT0.017270771S",
    "enqueuedAt": "2022-10-04T15:55:02.364638737Z",
    "startedAt": "2022-10-04T15:55:02.37723266Z",
    "finishedAt": "2022-10-04T15:55:02.394503431Z"
  },
  {
    "uid": 17,
    "indexUid": "dnd_spells",
    "status": "succeeded",
    "type": {
      "documentImport": {
        "method": "ReplaceDocuments",
        "allow_index_creation": true
      }
    },
    "details": {
      "receivedDocuments": 10,
      "indexedDocuments": 10
    },
    "duration": "PT0.017133299S",
    "enqueuedAt": "2022-10-04T15:55:02.116014762Z",
    "startedAt": "2022-10-04T15:55:02.128683566Z",
    "finishedAt": "2022-10-04T15:55:02.145816865Z"
  },
  {
    "uid": 16,
    "indexUid": "products",
    "status": "succeeded",
    "type": {
      "documentImport": {
        "method": "ReplaceDocuments",
        "allow_index_creation": true
      }
    },
    "details": {
      "receivedDocuments": 10,
      "indexedDocuments": 10
    },
    "duration": "PT0.017521995S",
    "enqueuedAt": "2022-10-04T15:55:01.867101853Z",
    "startedAt": "2022-10-04T15:55:01.879803378Z",
    "finishedAt": "2022-10-04T15:55:01.897325373Z"
  },
  {
    "uid": 15,
    "indexUid": "products",
    "status": "succeeded",
    "type": {
      "documentImport": {
        "method": "ReplaceDocuments",
        "allow_index_creation": true
      }
    },
    "details": {
      "receivedDocuments": 10,
      "indexedDocuments": 10
    },
    "duration": "PT0.011965881S",
    "enqueuedAt": "2022-10-04T15:55:01.245884922Z",
    "startedAt": "2022-10-04T15:55:01.258598227Z",
    "finishedAt": "2022-10-04T15:55:01.270564108Z"
  },
  {
    "uid": 14,
    "indexUid": "products",
    "status": "succeeded",
    "type": {
      "settings": {
        "allow_index_creation": true
      }
    },
    "details": {
      "synonyms": {
        "android": [
          "phone",
          "smartphone"
        ],
        "iphone": [
          "phone",
          "smartphone"
        ],
        "phone": [
          "smartphone",
          "iphone",
          "android"
        ]
      }
    },
    "duration": "PT0.010776512S",
    "enqueuedAt": "2022-10-04T15:55:00.630109164Z",
    "startedAt": "2022-10-04T15:55:00.643609011Z",
    "finishedAt": "2022-10-04T15:55:00.654385523Z"
  },
  {
    "uid": 13,
    "indexUid": "movies",
    "status": "succeeded",
    "type": {
      "settings": {
        "allow_index_creation": true
      }
    },
    "details": {
      "rankingRules": [
        "words",
        "typo",
        "proximity",
        "attribute",
        "sort",
        "exactness",
        "release_date:asc"
      ]
    },
    "duration": "PT0.010692540S",
    "enqueuedAt": "2022-10-04T15:55:00.412114406Z",
    "startedAt": "2022-10-04T15:55:00.425716285Z",
    "finishedAt": "2022-10-04T15:55:00.436408825Z"
  },
  {
    "uid": 12,
    "indexUid": "movies",
    "status": "succeeded",
    "type": {
      "settings": {
        "allow_index_creation": true
      }
    },
    "details": {
      "filterableAttributes": [
        "genres",
        "id"
      ],
      "sortableAttributes": [
        "release_date"
      ]
    },
    "duration": "PT0.010643910S",
    "enqueuedAt": "2022-10-04T15:55:00.18896188Z",
    "startedAt": "2022-10-04T15:55:00.207804798Z",
    "finishedAt": "2022-10-04T15:55:00.218448708Z"
  },
  {
    "uid": 11,
    "indexUid": "movies",
    "status": "succeeded",
    "type": {
      "documentImport": {
        "method": "ReplaceDocuments",
        "allow_index_creation": true
      }
    },
    "details": {
      "receivedDocuments": 10,
      "indexedDocuments": 10
    },
    "duration": "PT0.018142526S",
    "enqueuedAt": "2022-10-04T15:54:59.971297669Z",
    "startedAt": "2022-10-04T15:54:59.984799097Z",
    "finishedAt": "2022-10-04T15:55:00.002941623Z"
  },
  {
    "uid": 10,
    "indexUid": "movies",
    "status": "succeeded",
    "type": {
      "documentImport": {
        "method": "ReplaceDocuments",
        "allow_index_creation": true
      }
    },
    "details": {
      "receivedDocuments": 100,
      "indexedDocuments": 100
    },
    "duration": "PT0.366830299S",
    "enqueuedAt": "2022-10-04T15:51:44.147743385Z",
    "startedAt": "2022-10-04T15:51:44.458473756Z",
    "finishedAt": "2022-10-04T15:51:44.825304055Z"
  },
  {
    "uid": 9,
    "indexUid": "movies",
    "status": "succeeded",
    "type": {
      "documentImport": {
        "method": "ReplaceDocuments",
        "allow_index_creation": true
      }
    },
    "details": {
      "receivedDocuments": 90,
      "indexedDocuments": 90
    },
    "duration": "PT0.305133258S",
    "enqueuedAt": "2022-10-04T15:51:44.142779621Z",
    "startedAt": "2022-10-04T15:51:44.14805041Z",
    "finishedAt": "2022-10-04T15:51:44.453183668Z"
  },
  {
    "uid": 8,
    "indexUid": null,
    "status": "succeeded",
    "type": "dumpExport",
    "details": {
      "dumpUid": "20221004-155144042"
    },
    "duration": "PT0.022334966S",
    "enqueuedAt": "2022-10-04T15:51:44.042750953Z",
    "startedAt": "2022-10-04T15:51:44.056661399Z",
    "finishedAt": "2022-10-04T15:51:44.078996365Z"
  },
  {
    "uid": 7,
    "indexUid": "dnd_spells",
    "status": "succeeded",
    "type": {
      "documentImport": {
        "method": "ReplaceDocuments",
        "allow_index_creation": true
      }
    },
    "details": {
      "receivedDocuments": 10,
      "indexedDocuments": 10
    },
    "duration": "PT0.039310363S",
    "enqueuedAt": "2022-10-04T15:51:37.628596827Z",
    "startedAt": "2022-10-04T15:51:37.638650617Z",
    "finishedAt": "2022-10-04T15:51:37.67796098Z"
  },
  {
    "uid": 6,
    "indexUid": "dnd_spells",
    "status": "failed",
    "type": {
      "documentImport": {
        "method": "ReplaceDocuments",
        "allow_index_creation": true
      }
    },
    "details": {
      "receivedDocuments": 10,
      "indexedDocuments": 0
    },
    "error": {
      "message": "The primary key inference process failed because the engine did not find any fields containing `id` substring in their name. If your document identifier does not contain any `id` substring, you can set the primary key of the index.",
      "code": "unretrievable_error_code",
      "type": "invalid_request",
      "link": "https://docs.meilisearch.com/errors#unretrievable_error_code"
    },
    "duration": "PT0.128068051S",
    "enqueuedAt": "2022-10-04T15:51:37.381094632Z",
    "startedAt": "2022-10-04T15:51:37.394321835Z",
    "finishedAt": "2022-10-04T15:51:37.522389886Z"
  },
  {
    "uid": 5,
    "indexUid": "products",
    "status": "succeeded",
    "type": {
      "documentImport": {
        "method": "ReplaceDocuments",
        "allow_index_creation": true
      }
    },
    "details": {
      "receivedDocuments": 10,
      "indexedDocuments": 10
    },
    "duration": "PT0.021846888S",
    "enqueuedAt": "2022-10-04T15:51:37.132755272Z",
    "startedAt": "2022-10-04T15:51:37.146031377Z",
    "finishedAt": "2022-10-04T15:51:37.167878265Z"
  },
  {
    "uid": 4,
    "indexUid": "products",
    "status": "failed",
    "type": {
      "documentImport": {
        "method": "ReplaceDocuments",
        "allow_index_creation": true
      }
    },
    "details": {
      "receivedDocuments": 10,
      "indexedDocuments": 0
    },
    "error": {
      "message": "The primary key inference process failed because the engine did not find any fields containing `id` substring in their name. If your document identifier does not contain any `id` substring, you can set the primary key of the index.",
      "code": "unretrievable_error_code",
      "type": "invalid_request",
      "link": "https://docs.meilisearch.com/errors#unretrievable_error_code"
    },
    "duration": "PT0.005439591S",
    "enqueuedAt": "2022-10-04T15:51:36.53155275Z",
    "startedAt": "2022-10-04T15:51:36.544820074Z",
    "finishedAt": "2022-10-04T15:51:36.550259665Z"
  },
  {
    "uid": 3,
    "indexUid": "products",
    "status": "succeeded",
    "type": {
      "settings": {
        "allow_index_creation": true
      }
    },
    "details": {
      "synonyms": {
        "android": [
          "phone",
          "smartphone"
        ],
        "iphone": [
          "phone",
          "smartphone"
        ],
        "phone": [
          "smartphone",
          "iphone",
          "android"
        ]
      }
    },
    "duration": "PT0.135098240S",
    "enqueuedAt": "2022-10-04T15:51:35.939396731Z",
    "startedAt": "2022-10-04T15:51:35.952670314Z",
    "finishedAt": "2022-10-04T15:51:36.087768554Z"
  },
  {
    "uid": 2,
    "indexUid": "movies",
    "status": "succeeded",
    "type": {
      "settings": {
        "allow_index_creation": true
      }
    },
    "details": {
      "rankingRules": [
        "words",
        "typo",
        "proximity",
        "attribute",
        "sort",
        "exactness",
        "release_date:asc"
      ]
    },
    "duration": "PT0.010129921S",
    "enqueuedAt": "2022-10-04T15:51:35.721766758Z",
    "startedAt": "2022-10-04T15:51:35.735030412Z",
    "finishedAt": "2022-10-04T15:51:35.745160333Z"
  },
  {
    "uid": 1,
    "indexUid": "movies",
    "status": "succeeded",
    "type": {
      "settings": {
        "allow_index_creation": true
      }
    },
    "details": {
      "filterableAttributes": [
        "genres",
        "id"
      ],
      "sortableAttributes": [
        "release_date"
      ]
    },
    "duration": "PT0.022219445S",
    "enqueuedAt": "2022-10-04T15:51:35.50733827Z",
    "startedAt": "2022-10-04T15:51:35.517547002Z",
    "finishedAt": "2022-10-04T15:51:35.539766447Z"
  },
  {
    "uid": 0,
    "indexUid": "movies",
    "status": "succeeded",
    "type": {
      "documentImport": {
        "method": "ReplaceDocuments",
        "allow_index_creation": true
      }
    },
    "details": {
      "receivedDocuments": 10,
      "indexedDocuments": 10
    },
    "duration": "PT0.156616872S",
    "enqueuedAt": "2022-10-04T15:51:35.291992167Z",
    "startedAt": "2022-10-04T15:51:35.302593877Z",
    "finishedAt": "2022-10-04T15:51:35.459210749Z"
  }
]
